### PR TITLE
Security improvements

### PR DESCRIPTION
**Pull request by x676f64**
I have made some improvements to the contracts which will help prevent reentrancy vulnerabilities. I've also removed some extraneous payable declarations.

### CODE REVIEW

File: Contracts/contracts/Hub.sol [hunk 1]
```diff
@@ -172,25 +173,20 @@ contract Hub is ERC721, Ownable {
...
-
-        lock = false;
-
         if (msg.value > _deployPayment + _serverPayment) {
             _lostMoney[msg.sender] =
                 msg.value -
                 (_deployPayment + _serverPayment);
         }
-        return true;
+
```
Comment (Nethny) [inline Contracts/contracts/Hub.sol:184]: Why are you transferring the code for sending funds? This creates an opportunity for such a situation when the money was debited from the account, but the transfer itself was not executed for some reason
But it also works in the other direction, when it turned out to transfer the 1st time, but it was not possible to transfer to the next one, then you can write off as many funds as you want.
Funds should be debited after each transfer

**Review by Nethny — COMMENTED** [review]
Why are you transferring the code for sending funds?

File: Contracts/contracts/Hub.sol (hunk 1 above)
Comment (Nethny) [reply Contracts/contracts/Hub.sol:184]: In addition, in the takeMoney function, checking the balance will not allow you to re-enter this function, and extra checks are an extra waste of gas

Comment (x676f64) [reply Contracts/contracts/Hub.sol:184]: > Why are you transferring the code for sending funds? This creates an opportunity for such a situation when the money was debited from the account, but the transfer itself was not executed for some reason But it also works in the other direction, when it turned out to transfer the 1st time, but it was not possible to transfer to the next one, then you can write off as many funds as you want. Funds should be debited after each transfer

In the EVM, the entire call stack either succeeds or fails, there is no in between so if an external call fails and you have a `require(success)` all state changes revert. See the checks-effect-interactions to better understand why this is the correct implementation. You do not want to make state changes after external calls.

> In addition, in the takeMoney function, checking the balance will not allow you to re-enter this function, and extra checks are an extra waste of gas
https://github.com/rmrk-team/remov/blob/5da294860481173e3de0ec5c4a0a484d3e4ed037/Contracts/contracts/Hub.sol#L170

This a classic example of a reentrancy vulnerability. While at the moment, I don't think it has an economically exploitable impact, we are building money legos here which could have implications later on. You should have a reentrancyGuard implemented on any function that makes external calls to untrusted sources.

**Pull request merged**